tests/util.py

```diff
@@ -0,0 +1,171 @@
+        return query
+
+
+def read(*paths: str, decoder: Optional[Callable] = None, **kwargs):
+    with open(os.path.join(*paths), **kwargs) as f:
+        return decoder(f) if decoder else f.read()
+
+
+def ndjson_load(file_obj):
+    return [json.loads(line) for line in file_obj]
+
+
+def load(resource_dir: str, basename: str, **search):
+    search = search or {"yaml": yaml.load, "json": json.load, "ndjson": ndjson_load}
+    for ext, decoder in search.items():
+        try:
+            return read(resource_dir, f"{basename}.{ext}", decoder=decoder)
+        except FileNotFoundError:
+            pass
+    not_found = "', '".join(map(lambda ext: f"{basename}.{ext}", search))
+    raise FileNotFoundError(f"[Errno 2] No such file or directory: '{not_found}'")
+
+
+def get_query_params(resource_dir: str):
+    try:
+        params = load(resource_dir, "query_params")
+    except FileNotFoundError:
+        params = []
+    for param in params:
+        if {"name", "type", "type_", "value"}.issuperset(param.keys()):
+            # this is a scalar query param
+            param["type_"] = param.pop("type", param.pop("type_", "STRING"))
+            yield bigquery.ScalarQueryParameter(**param)
+        else:
+            # attempt to coerce to some type of query param
+            try:
+                yield bigquery.StructQueryParameter.from_api_repr(param)
+            except KeyError:
+                try:
+                    yield bigquery.ArrayQueryParameter.from_api_repr(param)
+                except KeyError:
+                    # this is a different format for scalar param than above
+                    yield bigquery.ScalarQueryParameter.from_api_repr(param)
+
+
+def generate_tests():
+    tests_dir = os.path.dirname(__file__)
+    sql_dir = os.path.join(os.path.dirname(tests_dir), "sql")
+
+    # iterate over directories in tests_dir
+    for query_name in next(os.walk(tests_dir))[1]:
+        query_dir = os.path.join(tests_dir, query_name)
+
+        # read query or skip
+        try:
+            query = read(sql_dir, f"{query_name}.sql")
+        except FileNotFoundError:
+            continue
+
+        # generate a test for each directory in query_dir
+        for test_name in next(os.walk(query_dir))[1]:
+            resource_dir = os.path.join(query_dir, test_name)
+            query_params = get_query_params(resource_dir)
+            tables: Set[Table] = set()
+            replace: Dict[str, str] = {}
+
+            # load expect or skip
+            try:
+                expect = load(resource_dir, "expect")
+            except FileNotFoundError:
+                continue
+
+            # generate tables for files with a supported table extension
+            for resource in next(os.walk(resource_dir))[2]:
+                if resource in ("expect.ndjson", "query_params.ndjson"):
+                    continue  # expect and query_params are not tables
+                table_name, extension = resource, ""
```

Comment:
this should move above line 130, and line 130 should use `table_name` instead of `resource`